modules/core/src/vaapi.cpp

```diff
@@ -173,18 +178,161 @@ static bool ocl_convert_bgr_to_nv12(cl_mem clBuffer, int step, int cols, int row
 
 } // namespace cv::vaapi::ocl
 
-void convertToVASurface(InputArray src, VASurfaceID surface, Size size)
+#if defined(HAVE_VA)
+const int NCHANNELS = 3;
+
+static void copy_convert_nv12_to_bgr(const VAImage& image, const unsigned char* buffer, Mat& bgr)
+{
```

Comment:
looks like this code overlaps with imgproc one (but this one should be much slow). The recommended way is to port imgproc's impl to HAL and reuse it here